# Desafío: DATA WRANGLING

**Consigna:** Leer los set de datos w_data.csv y p_data.csv, fusionarlos a partir de la variable que tengan en común(ID) y realizar las siguientes ejercicios:
1. Cuantos ID tienen en común. Nombrar cuales no tienen en común
2. Corregir los target a solo dos valores distintos ‘<=50k’ y ‘>50k’
3. Calcular la proporción que hay entre personas con salarios >50k respecto del total para personas de distinto sexo, Raza y Educacion. Por ejemplo cuál es el porcentaje de mujeres con un salario mayor a 50K y compararlo con el de los hombres. Lo mismo para personas de distintas razas y Educación

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df_w = pd.read_csv('C:/Users/Usuario/Documents/Cursos/Coderhouse/files/w_data.csv')
df_p = pd.read_csv('C:/Users/Usuario/Documents/Cursos/Coderhouse/files/p_data.csv')

In [ ]:
df_w

Workclass;fnlwgt;Capital Gain;Capital Loss;Hours per week;ID;Target
0         State-gov;77516.0;2174.0;0.0;40.0;49084; <=50K                 
1       Self-emp-not-inc;83311.0;0.0;0.0;13.0;68553; ...                 
2            Private;215646.0;0.0;0.0;40.0;143654; <=50K                 
3             Private;234721.0;0.0;0.0;40.0;41510; <=50K                 
4            Private;338409.0;0.0;0.0;40.0;195025; <=50K                 
...                                                  ...                 
48838    Private;215419.0;0.0;0.0;36.0;735617153; <=50K.                 
48839         ?;321403.0;0.0;0.0;40.0;1131807446; <=50K.                 
48840    Private;374983.0;0.0;0.0;50.0;781422087; <=50K.                 
48841   Private;83891.0;5455.0;0.0;40.0;161771296; <=...                 
48842   Self-emp-inc;182148.0;0.0;0.0;60.0;1344995623...                 

[48843 rows x 1 columns]

In [ ]:
df_w.columns

Index(['Workclass;fnlwgt;Capital Gain;Capital Loss;Hours per week;ID;Target'], dtype='object')

In [ ]:
df_w_columns = pd.Series(df_w['Workclass;fnlwgt;Capital Gain;Capital Loss;Hours per week;ID;Target'])
df_w_columns= df_w_columns.str.split(';', expand=True)
df_w_columns.columns = ['Workclass','fnlwgt','Capital Gain','Capital Loss','Hours per week','ID','Target']

In [ ]:
df_w_columns

Workclass    fnlwgt Capital Gain Capital Loss Hours per week  \
0              State-gov   77516.0       2174.0          0.0           40.0   
1       Self-emp-not-inc   83311.0          0.0          0.0           13.0   
2                Private  215646.0          0.0          0.0           40.0   
3                Private  234721.0          0.0          0.0           40.0   
4                Private  338409.0          0.0          0.0           40.0   
...                  ...       ...          ...          ...            ...   
48838            Private  215419.0          0.0          0.0           36.0   
48839                  ?  321403.0          0.0          0.0           40.0   
48840            Private  374983.0          0.0          0.0           50.0   
48841            Private   83891.0       5455.0          0.0           40.0   
48842       Self-emp-inc  182148.0          0.0          0.0           60.0   

               ID   Target  
0           49084    <=50K  
1           68553    <=50K  
2          143654    <=50K  
3           41510    <=50K  
4          195025    <=50K  
...           ...      ...  
48838   735617153   <=50K.  
48839  1131807446   <=50K.  
48840   781422087   <=50K.  
48841   161771296   <=50K.  
48842  1344995623    >50K.  

[48843 rows x 7 columns]

In [ ]:
df_p

Age;Sex;ID;Race;Country;Education;Education-Num;Marital Status;Relationship
0      39; Male;49084; White; United-States; Bachelor...                         
1      50; Male;68553; White; United-States; Bachelor...                         
2      38; Male;143654; White; United-States; HS-grad...                         
3      53; Male;41510; Black; United-States; 11th;7.0...                         
4      28; Female;195025; Black; Cuba; Bachelors;13.0...                         
...                                                  ...                         
48837  33; Male;597186815; White; United-States; Bach...                         
48838  39; Female;735617153; White; United-States; Ba...                         
48839  64; Male;1131807446; Black; United-States; HS-...                         
48840  38; Male;781422087; White; United-States; Bach...                         
48841  44; Male;161771296; Asian-Pac-Islander; United...                         

[48842 rows x 1 columns]

In [ ]:
df_p.columns

Index(['Age;Sex;ID;Race;Country;Education;Education-Num;Marital Status;Relationship'], dtype='object')

In [ ]:
df_p_columns = pd.Series(df_p['Age;Sex;ID;Race;Country;Education;Education-Num;Marital Status;Relationship'])
df_p_columns= df_p_columns.str.split(';', expand=True)
df_p_columns.columns = ['Age','Sex','ID','Race','Country','Education','Education-Num','Marital Status','Relationship']

In [ ]:
df_p_columns

Age      Sex          ID                 Race         Country  \
0      39     Male       49084                White   United-States   
1      50     Male       68553                White   United-States   
2      38     Male      143654                White   United-States   
3      53     Male       41510                Black   United-States   
4      28   Female      195025                Black            Cuba   
...    ..      ...         ...                  ...             ...   
48837  33     Male   597186815                White   United-States   
48838  39   Female   735617153                White   United-States   
48839  64     Male  1131807446                Black   United-States   
48840  38     Male   781422087                White   United-States   
48841  44     Male   161771296   Asian-Pac-Islander   United-States   

        Education Education-Num       Marital Status     Relationship  
0       Bachelors          13.0        Never-married    Not-in-family  
1       Bachelors          13.0   Married-civ-spouse          Husband  
2         HS-grad           9.0             Divorced    Not-in-family  
3            11th           7.0   Married-civ-spouse          Husband  
4       Bachelors          13.0   Married-civ-spouse             Wife  
...           ...           ...                  ...              ...  
48837   Bachelors          13.0        Never-married        Own-child  
48838   Bachelors          13.0             Divorced    Not-in-family  
48839     HS-grad           9.0              Widowed   Other-relative  
48840   Bachelors          13.0   Married-civ-spouse          Husband  
48841   Bachelors          13.0             Divorced        Own-child  

[48842 rows x 9 columns]

In [ ]:
df = pd.merge(df_w_columns, df_p_columns, on='ID', how = 'outer', indicator = True)
df

Workclass    fnlwgt Capital Gain Capital Loss Hours per week  \
0              State-gov   77516.0       2174.0          0.0           40.0   
1       Self-emp-not-inc   83311.0          0.0          0.0           13.0   
2                Private  215646.0          0.0          0.0           40.0   
3                Private  234721.0          0.0          0.0           40.0   
4                Private  338409.0          0.0          0.0           40.0   
...                  ...       ...          ...          ...            ...   
48838            Private  215419.0          0.0          0.0           36.0   
48839                  ?  321403.0          0.0          0.0           40.0   
48840            Private  374983.0          0.0          0.0           50.0   
48841            Private   83891.0       5455.0          0.0           40.0   
48842       Self-emp-inc  182148.0          0.0          0.0           60.0   

               ID   Target  Age      Sex                 Race         Country  \
0           49084    <=50K   39     Male                White   United-States   
1           68553    <=50K   50     Male                White   United-States   
2          143654    <=50K   38     Male                White   United-States   
3           41510    <=50K   53     Male                Black   United-States   
4          195025    <=50K   28   Female                Black            Cuba   
...           ...      ...  ...      ...                  ...             ...   
48838   735617153   <=50K.   39   Female                White   United-States   
48839  1131807446   <=50K.   64     Male                Black   United-States   
48840   781422087   <=50K.   38     Male                White   United-States   
48841   161771296   <=50K.   44     Male   Asian-Pac-Islander   United-States   
48842  1344995623    >50K.  NaN      NaN                  NaN             NaN   

        Education Education-Num       Marital Status     Relationship  \
0       Bachelors          13.0        Never-married    Not-in-family   
1       Bachelors          13.0   Married-civ-spouse          Husband   
2         HS-grad           9.0             Divorced    Not-in-family   
3            11th           7.0   Married-civ-spouse          Husband   
4       Bachelors          13.0   Married-civ-spouse             Wife   
...           ...           ...                  ...              ...   
48838   Bachelors          13.0             Divorced    Not-in-family   
48839     HS-grad           9.0              Widowed   Other-relative   
48840   Bachelors          13.0   Married-civ-spouse          Husband   
48841   Bachelors          13.0             Divorced        Own-child   
48842         NaN           NaN                  NaN              NaN   

          _merge  
0           both  
1           both  
2           both  
3           both  
4           both  
...          ...  
48838       both  
48839       both  
48840       both  
48841       both  
48842  left_only  

[48843 rows x 16 columns]

In [ ]:
df['ID'].groupby(df['_merge']).count()

_merge
left_only         1
right_only        0
both          48842
Name: ID, dtype: int64

**Respuesta Ejercicio 1:**

Los ID en común entre los datasets suman en total: 48842

El ID distinto es el 1344995623

**Ejercicio 2**

In [ ]:
df['Target'].unique()

array([' <=50K', ' >50K', '', ' <=50K.', ' >50K.'], dtype=object)

In [ ]:
df.replace({' <=50K.':' <=50K',' >50K.':' >50K'}, inplace=True)
df['Target'].unique()

array([' <=50K', ' >50K', ''], dtype=object)

Busco el registro con target ' ' y lo borro al estar vacío

In [ ]:
df[df['Target']=='']

Workclass fnlwgt Capital Gain Capital Loss Hours per week     ID Target  \
32561                                                            37078          

      Age Sex Race Country Education Education-Num Marital Status  \
32561                                                               

      Relationship _merge  
32561                both

In [ ]:
df.drop(labels = 32561, inplace=True)
df['Target'].unique()

array([' <=50K', ' >50K'], dtype=object)

**Ejercicio 3**

In [ ]:
cant_registros = 48842

def porcentaje(n):
  return (n/cant_registros) * 100

Proporción de salarios respecto del total para personas de distinto sexo

In [ ]:
df['Sex'].groupby(df['Target']).value_counts(dropna=False).apply(porcentaje)

Target  Sex    
 <=50K   Male      46.541911
         Female    29.529913
 >50K    Male      20.304246
         Female     3.621883
        NaN         0.002047
Name: Sex, dtype: float64

Proporción de salarios respecto del total para personas de distinta raza

In [ ]:
df['Race'].groupby(df['Target']).value_counts(dropna=False).apply(porcentaje)

Target  Race               
 <=50K   White                 63.787314
         Black                  8.433316
         Asian-Pac-Islander     2.272634
         Amer-Indian-Eskimo     0.849679
         Other                  0.728881
 >50K    White                 21.714917
         Black                  1.158839
         Asian-Pac-Islander     0.837394
         Amer-Indian-Eskimo     0.112608
         Other                  0.102371
        NaN                     0.002047
Name: Race, dtype: float64

Proporción de salarios respecto del total para personas de distinta educación

In [ ]:
df['Education'].groupby(df['Target']).value_counts(dropna=False).apply(porcentaje)

Target  Education    
 <=50K   HS-grad         27.191761
         Some-college    18.047991
         Bachelors        9.647435
         11th             3.521559
         Assoc-voc        3.150977
         10th             2.665739
         Masters          2.452807
         Assoc-acdm       2.432333
         7th-8th          1.828344
         9th              1.463904
         12th             1.246878
         5th-6th          0.986856
         1st-4th          0.489333
         Prof-school      0.444290
         Doctorate        0.333729
         Preschool        0.167888
 >50K    Bachelors        6.781049
         HS-grad          5.124688
         Some-college     4.223824
         Masters          2.987183
         Prof-school      1.263257
         Assoc-voc        1.068752
         Doctorate        0.882437
         Assoc-acdm       0.845584
         11th             0.188362
         10th             0.178125
         7th-8th          0.126940
         12th             0.09827